blockchain/consensus/narwhal_cns/src/proposer.rs

```diff
@@ -0,0 +1,420 @@
+    keypair: KeyPair,
+    committee: Committee,
+    validator_addresses: HashMap<PublicKey, Address>,
+    storage_base_path: PathBuf,
+    parameters: Parameters,
+}
+
+#[async_trait]
+impl Proposer for NarwhalProposer {
+    /// Start a Narwhal Primary and a Worker in the background and subscribe to
+    /// local chain extensions. Whenever there is a new tip: (1) check for new
+    /// transactions that can be sent to Narwhal and (2) take the next output
+    /// from Narwhal and turn it into the next block.
+    async fn spawn<DB, MP>(
+        self,
+        state_manager: Arc<StateManager<DB>>,
+        mpool: Arc<MP>,
+        submitter: SyncGossipSubmitter,
+    ) -> anyhow::Result<Vec<JoinHandle<()>>>
+    where
+        DB: BlockStore + Sync + Send + 'static,
+        MP: MessagePoolApi + Send + Sync + 'static,
+    {
+        let chain_store = state_manager.chain_store();
+        let (output_tx, output_rx) = channel::bounded(self.committee.size());
+        let (head_tx, head_rx) = tokio::sync::watch::channel(None);
+        let execution_state = Arc::new(NarwhalExecutionState::new(chain_store.clone(), output_tx));
+
+        // The following is based on the `NodeRestarter` in the narwhal repo.
+
+        // Get a store for the epoch of the committee (although currently there won't be reconfiguration.)
+        let mut store_path = self.storage_base_path.clone();
+        store_path.push(format!("epoch{}", self.committee.epoch()));
+        let store = NodeStorage::reopen(store_path);
+
+        let registry = prometheus::Registry::default();
+        let name = self.keypair.public().clone();
+
+        // Lazily connect to the worker which isn't running yet.
+        let transactions_client = make_transactions_client(&name, &self.committee)?;
+
+        // Start the Primary and a Worker. They will ask the execution state where to resume from.
+        let mut handles = Vec::new();
+
+        let primary_handles = Node::spawn_primary(
+            self.keypair,
+            Arc::new(ArcSwap::new(Arc::new(self.committee.clone()))),
+            &store,
+            self.parameters.clone(),
+            true,
+            execution_state,
+            &registry,
+        )
+        .await?;
+
+        let worker_handles = Node::spawn_workers(
+            name,
+            vec![0],
+            Arc::new(ArcSwap::new(Arc::new(self.committee))),
+            &store,
+            self.parameters,
+            &registry,
+        );
+
+        handles.extend(tokio_to_async_std(primary_handles));
+        handles.extend(tokio_to_async_std(worker_handles));
+
+        // Subscribe to local chain extensions.
+        let head_change_rx = chain_store.publisher().subscribe();
+
+        // Get the current head of the chain.
+        let current_head = chain_store.heaviest_tipset().await;
+
+        // NOTE: Instead of passig in the current head, we could use `chain_store.sub_header_changes` and `chain_store.next_header_change`,
```

Comment:
```suggestion
        // NOTE: Instead of passing in the current head, we could use `chain_store.sub_header_changes` and `chain_store.next_header_change`,
```